lib/manager/gradle/shallow/extract.ts

```diff
@@ -64,6 +80,70 @@ export async function extractAllPackageFiles(
+        // Implement TOML file parsing and extraction
+        const tomlContent = parse(content) as GradleCatalog;
+        const versions = tomlContent.versions || {};
+        const libs = tomlContent.libraries || {};
+        const libStartIndex = content.indexOf('libraries');
+        const libSubContent = content.slice(libStartIndex);
+        const versionStartIndex = content.indexOf('versions');
+        const versionSubContent = content.slice(versionStartIndex);
+        for (const libraryName of Object.keys(libs)) {
+          const libDescriptor = libs[libraryName];
+          const group: string =
+            typeof libDescriptor === 'string'
+              ? libDescriptor.split(':')[0]
+              : libDescriptor.group || libDescriptor.module?.split(':')[0];
+          const name: string =
+            typeof libDescriptor === 'string'
+              ? libDescriptor.split(':')[1]
+              : libDescriptor.name || libDescriptor.module?.split(':')[1];
+          const version = libDescriptor.version || libDescriptor.split(':')[2];
+          const currentVersion =
+            typeof version === 'string' ? version : versions[version.ref];
+          const fileReplacePosition =
+            typeof version === 'string'
+              ? libStartIndex +
+                findIndexAfter(libSubContent, libraryName, currentVersion)
+              : versionStartIndex +
+                findIndexAfter(versionSubContent, version.ref, currentVersion);
+          const dependency = {
+            depName: `${group}:${name}`,
+            groupName: group,
+            currentValue: currentVersion,
+            managerData: { fileReplacePosition, packageFile },
+          };
+          extractedDeps.push(dependency);
+        }
+        const plugins = tomlContent.plugins || {};
+        const pluginsStartIndex = content.indexOf('[plugins]');
+        const pluginsSubContent = content.slice(pluginsStartIndex);
+        for (const pluginName of Object.keys(plugins)) {
+          const pluginDescriptor = plugins[
+            pluginName
+          ] as GradleCatalogPluginDescriptor;
+          const pluginId = pluginDescriptor.id;
+          const version = pluginDescriptor.version;
+          const currentVersion: string =
+            typeof version === 'string' ? version : versions[version.ref];
+          const fileReplacePosition =
+            typeof version === 'string'
+              ? pluginsStartIndex +
+                findIndexAfter(pluginsSubContent, pluginId, currentVersion)
+              : versionStartIndex +
+                findIndexAfter(versionSubContent, version.ref, currentVersion);
+          const dependency = {
+            depType: 'plugin',
+            depName: pluginId,
+            lookupName: `${pluginId}:${pluginId}.gradle.plugin`,
+            registryUrls: ['https://plugins.gradle.org/m2/'],
+            currentValue: currentVersion,
+            commitMessageTopic: `plugin ${pluginName}`,
```

Comment:
Remove
```suggestion
```

Checkout terraform manager for how to override this

Reply:
I looked at how it's implemented here: https://github.com/renovatebot/renovate/blob/715933850929f409f0cfad07cee159a60091414d/lib/manager/gradle/shallow/parser.ts#L184

It looks similar to me, what is wrong? I looked at how it's done in Terraform, and I do not think it achieves the same: I want the topic to be different for plugins only, while there the change seems to be applied everywhere.

Can you help me?

Reply:
@viceice I tested this also with a true repository, and the result seems correct. Removing that line confuses plugins and libraries with the same name.

Reply:
ok, but can you extract the big new conditional block to a new function?

Best would be to a new file `./extract/catalog.ts` or something like this.

I'm a little concerned about the huge function, we can extract the other blocks in a later pr.

Reply:
Just in case, please wrap extracted function with try-catch block. It would be great to have `logger.warn('TOML parsing error', { err })` in catch block.

Reply:
@viceice done

Reply:
@zharinov done